### ISSUE TITLE
bug: RecursionError when trying to register local directory of partitioned parquet files

### ISSUE THREAD

**Issue by markhng525** [ibis-project/ibis#4981]
### What happened?

Hi all I'm experiencing an issue with trying to register a directory of partitioned parquet files as an ibis table.  Using just `duckdb` it works perfectly fine
```
import duckdb
import pyarrow.dataset
import pyarrow.fs
from pyarrow.parquet import ParquetDataset

arrow_dataset: ParquetDataset = pyarrow.dataset.dataset("/Users/my/Data", filesystem=pyarrow.fs.LocalFileSystem(), format="parquet")
duckdb_conn = duckdb.connect()
duckdb_conn.execute("select count(1) from arrow_dataset")
```

However, when I try to register the data source using `ibis` it throws an error.
```
import ibis
from ibis.backends.duckdb import Backend as DuckDBBackend

ibis_conn: DuckDBBackend = ibis.duckdb.connect(":memory:")
ibis_conn.register(
source="Users/my/Data",
table_name="events",
)
...
---> 69 it = reversed(parts)
70 for part in it:
71 if not part:

RecursionError: maximum recursion depth exceeded while calling a Python object
```

### What version of ibis are you using?

ibis-framework = {extras = ["duckdb"], version = "3.2.0"}

### What backend(s) are you using, if any?

duckdb = "0.6.0"

### Relevant log output

```sh
RecursionError                            Traceback (most recent call last)
Cell In[37], line 5
      2 from ibis.backends.duckdb import Backend as DuckDBBackend
      4 ibis_conn: DuckDBBackend = ibis.duckdb.connect(":memory:")
----> 5 ibis_conn.register(
      6     source="Users/my/Data",
      7     table_name="events",
      8 )

File ~/.venv/lib/python3.8/site-packages/ibis/backends/duckdb/__init__.py:174, in Backend.register(self, source, table_name)
    155 """Register a data source as a table in the current database.
    156 
    157 Parameters
   (...)
    171     The just-registered table
    172 """
    173 if isinstance(source, (str, Path)):
--> 174     sql, table_name = _generate_view_code(
    175         source, table_name=table_name
    176     )
    177     self.con.execute(sql)
    178 else:

File ~/.venv/lib/python3.8/site-packages/ibis/common/dispatch.py:104, in RegexDispatcher.__call__(self, s, *args, **kwargs)
...
---> 69 it = reversed(parts)
     70 for part in it:
     71     if not part:

RecursionError: maximum recursion depth exceeded while calling a Python object
```

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by gforsyth** [comment]
Hey @markhng525 -- thanks for reporting!  
I think the recursion error part of this has already been fixed in #4922 

But as for the broader failure mode here, we dispatch on either file extension type, or using URIs, so you should be able to do what you want with either:

```
con.register("/Users/my/Data/*.parquet", table_name="events")
```
(if the individual parquet files have the `.parquet` extension)

or

```
con.register("parquet:///Users/my/Data/*", table_name="events")
```

**Comment by markhng525** [comment]
@gforsyth that solved it!
